### ISSUE TITLE
Let apps re-use our bundled SQLite for their own storage needs

### ISSUE THREAD

**Issue by data-sync-user** [mozilla/uniffi-rs#953]
This is a follow-up to https://github.com/mozilla/application-services/pull/740, which came about from Firefox for iOS accidentally discovering that their apps own uses of SQLite were linking against the copy bundled in the app-services megazord.

We should expect and support client applications to use SQLite for storing other data for their own purposes, and it would be nice if they didn't have to bundle a second copy of SQLite in order to do so.  Let's figure out how to allow them to re-use ours (deliberately rather than by accident).

There are a few ways we could do this:

* Encourage linking against the sqlite symbols in our megazord. This is probably simplest from our PoV, but I would worry about unintended consequences or backwards-compatibility concerns when we want to make changes to sqlite for our own purposes.
* Expose a low-level sqlite API component for creating new database files. We could implement some amount of clever connection management that the app gets for free, and have more control over the linkage and API compatibility and what-not, but apps would write their data into a completely separate file than our components.
* Expose a higher-level sqlite API component for manipulating data inside a shared database file. On Firefox for iOS right now, IIRC all the app's data is stored in a single "browser.db" file.  Maybe we actually want to do something similar long term, and should figure out how to support side-by-side data managed by the app and by our components?

I don't yet have any strong opinions on the right path forward, but wanted to get this on file.  /cc @garvankeeley @justindarc @thomcc

**Comment by data-sync-user** [comment]
➤ Thom Chiovoloni commented:

Worth noting that on android bundling an additional SQLite is not needed, they can use https://developer.android.com/reference/android/database/sqlite/package-summary.html

Number 1 is only really possible on iOS, where we absolutely should do it. On android we might be able to make it work, but it would actually require doing 2 in practice as well if we did this, since any code that wants to use SQLite on android will be using some library (such as android.database.sqlite) which is almost certainly not going to let us pass in a jna.Library it can use to access the SQLite symbols (it would probably be using JNI to call more directly, which isn't possible unless that's compiled into us).

I think 2 would be a deceptively large amount of work, for very little benefit, and I doubt we'd have the time / resources to make it on-par with SQLite bindings for the platform.

Which leaves 3, which I'm not opposed to, and seems like it fits in with the kinds of things we tend to provide. That said, it depends what this looks like. I'd be against anything too generic, since it be complex to expose over the FFI.

To elaborate a bit, my hesitation with 2/3 (and by extension 1 on android) is basically because the FFI forces us to provide a C-compatible interface. SQLite already provides a C-compatible interface. If we just are providing bindings to SQLite, we've basically introduced a bunch of overhead, and probably written a buggier API than SQLite (since we don't have SQLite's literally nine million lines of test code to lean on).

**Comment by data-sync-user** [comment]
➤ Garvan Keeley commented:

Use-cases:
1) App needs SQLite and doesn't have a requirement for a SQLite with different compilation flags set.
Solution: The static iOS megazord exposes the bundled SQLite as-is, so yay.

2) App needs SQLite with a different set of compilation flags
Solution # 1: Remove SQLite binary from the static megazord, client app can link against custom SQLite, client app would need to ensure the compilation flags meet the requirements of the megazord.
Solution # 2: Do what Fx iOS does currently, and link a second SQLite. This is a worst-case scenario (near-duplicate binaries, tricky linking to ensure correct symbols are used, have fun figuring out which SQLite you are debugging), but is temporary because the code is transitioning from old storage system to new.

Case 1 above is the one I support –and it works as-is. The second case seems exceedingly rare. 

However, an additional missed use-case here could throw a spanner in the works.

**Comment by data-sync-user** [comment]
➤ Justin D&#39;Arcangelo commented:

I agree with thomcc that exposing a low-or-high level SQLite API from the megazord with things like connection management, etc. is a rabbit hole that we probably don't want to go down. It has taken a long time to get our DB code in iOS in a good state where it is now and doing so would effectively be like replicating months of that same work, but in a cross-platform manner.

Also, I second what garvankeeley is saying. That another app would need this level of control over SQLite to squeeze every ounce of performance out of it would seem rather unlikely. The entire reason for this requirement when building Firefox for iOS has to do with the fact that some users have an abnormally-large amount of data (history, etc.) and the app has been fine-tuned over time to optimize performance assuming a very-specific set of SQLite compilation options. I would imagine that most other mobile apps that require a SQLite database would just assume they're using the default system OS-provided library. FWIW, the current compilation options being used by the megazord seem to match nearly identical to what iOS ships by default.

**Comment by data-sync-user** [comment]
➤ Justin D&#39;Arcangelo commented:

tl;dr: I would prefer that we simply encourage apps to link to the SQLite bundled with the megazord and resort to the Firefox for iOS fallback method only when absolutely necessary.

The main action item I think we have now is to find an optimized set of compilation options we can use for SQLite in the megazord. If we can identify the options that result in the most performance, it is likely a desirable outcome for everyone.

**Comment by data-sync-user** [comment]
➤ Thom Chiovoloni commented:

I think we should also consider disabling cipher_memory_security in more cases if we want this to be usable as a replacement for SQLite. It causes SQLcipher to call mlock on every call to malloc, and munlock on every call to free. This is outrageously expensive on many platforms (macOS, iOS, probably others). And has a dubious security benefit [0]. 

Sadly, it's a global option, so it can't be adjusted only for connections that care about security (although, places acts like it is ( https://github.com/mozilla/application-services/blob/4fcb2078774ecd1f963e3ac2c8061e2305701c97/components/places/src/db/db.rs#L61 ), since the fact that this setting was global wasn't clear to me when writing that code).

[0]: mlock and munlock adjust whether or not pages are allowed to be swapped to disk. When properly used, this can prevent situations where keys are written to disk in cleartext by the operating system as part of the swap file.

However, these functions operate on units of entire pages, but the use in sqlcipher treats them as if they operate on units of bytes (see sqlcipher_mem_malloc and sqlcipher_mem_free in https://github.com/sqlcipher/sqlcipher/blob/fda4c68bb474da7e955be07a2b807bda1bb19bd2/src/crypto_impl.c). This mismatch means it's extremely possible (likely, even) for a page to be unlocked while it has sensitive data.

For example: if two allocations are performed, whose total size is less than a single page, and that share the same page, if one of them is freed but not the other, then the one that is not yet freed will no longer be protected.

Additionally, if we actually want to defend against this attack, we need to be zeroing memory when it's released as well (note that SQLcipher does perform this if that pragma is turned on). We can do this pretty easily in Rust (https://github.com/mozilla/application-services/issues/699 has a link to a branch where I implement it, although I have not actually guaranteed that the compiler has not removed the attempt to zero the memory, just made an attempt to ensure it), but there's no way to do this on Android, and probably none on iOS either.

That said, it's not that this is worthless. They'd help Lockbox, certainly, just that they're not worth the 6-8x perf hit I've seen.

**Comment by data-sync-user** [comment]
➤ Justin D&#39;Arcangelo commented:

100% agree to turn off cipher_memory_security by default. If Lockbox wants to use it, we can expose an API for them to turn it back on.
